### PR TITLE
add: add logic to send new user data to web3 and get wallet_token (#415)

### PR DESCRIPTION
**Pull request by yaroslav-fedyshyn-nordwhale**
This PR closes #415 #455
 * W3 user login token is taken from W3 after a successful registration and saved
 * save login token

 * add api route to fetch login token

PR for dapp changes - https://github.com/GoodDollar/GoodDAPP/pull/487

### CODE REVIEW

File: src/server/storage/storageAPI.js [hunk 1]
```diff
@@ -38,16 +41,45 @@ const setup = (app: Router, storage: StorageAPI) => {
...
+              full_name: user.fullName,
+              wallet_address: walletAddress
+            }
+          }).then(res => res.json())
+
+          if (web3Response.wallet_token) {
+            updateUserObj.w3Token = web3Response.wallet_token
+          }
+        } catch (e) {
+          log.debug('Get Web3 Login Response Failed', e)
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: log.error

File: src/server/storage/storageAPI.js [hunk 2]
```diff
@@ -38,16 +41,45 @@ const setup = (app: Router, storage: StorageAPI) => {
...
+
+      const updateUserObj = {
+        ...user,
+        mauticId
+      }
+
+      if (!user.w3Token) {
+        try {
+          const secureHash = md5(`${user.email}${conf.secure_key}`)
+          const web3Response = await fetch(`${conf.web3SiteUrl}/api/wl/user`, {
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: use Promise.all
combine with mautic.createContact
mauticRecordPromise = development? Promise.resolve({}) : Mautic.create...catch...
w3Promise = fetch...catch...
[mauticRecord,we3Record] = await Promise.all(...)

File: src/server/storage/storageAPI.js [hunk 3]
```diff
@@ -38,16 +41,45 @@ const setup = (app: Router, storage: StorageAPI) => {
...
       const mauticRecord = process.env.NODE_ENV === 'development' ? {} : await Mautic.createContact(user).catch(e => {})
       const mauticId = get(mauticRecord, 'contact.fields.all.id', -1)
       logger.debug('User mautic record', { mauticId, mauticRecord })
+
+      const updateUserObj = {
+        ...user,
+        mauticId
+      }
+
+      if (!user.w3Token) {
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: this is going to be false also for users that did came from W3.
if the api on W3 will work ok also for existing emails that's fine... will it?

File: src/server/storage/storageAPI.js [hunk 4]
```diff
@@ -38,16 +41,45 @@ const setup = (app: Router, storage: StorageAPI) => {
...
+        } catch (e) {
+          log.debug('Get Web3 Login Response Failed', e)
+        }
+      }
+
+      storage.updateUser(updateUserObj)
+
       //topwallet of user after registration
-      storage.updateUser({ ...user, mauticId })
       let ok = await AdminWallet.topWallet(userRecord.gdAddress, null, true)
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: I think we need to return the W3 token to the user, to be saved on his profile (encrypted)
So on the UI we can create the link to w3.gooddollar.org with the login token

File: src/server/storage/storageAPI.js (hunk 3 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: You mean that here is needed to check if such email already exists on web3?

File: src/server/storage/storageAPI.js (hunk 1 above)
Comment (sirpy) [reply src/server/storage/storageAPI.js]: @yaroslav-fedyshyn-nordwhale

File: .env.example [hunk 5]
```diff
@@ -37,4 +37,6 @@ SKIP_EMAIL_VERIFICATION=true
 SKIP_FACE_RECO=false
 ALLOW_FACE_RECO_DUPS=false
 ## Token for rollbar logging service
-ROLLBAR_TOKEN=
\ No newline at end of file
+ROLLBAR_TOKEN=
+WEB3_SITE_URL=https://w3.gooddollar.org
```
Comment (sirpy) [inline .env.example]: no values .env.example

File: .env.example [hunk 6]
```diff
@@ -37,4 +37,6 @@ SKIP_EMAIL_VERIFICATION=true
...
+SECURE_KEY=d41d8cd98f00b204e9800998ecf8427e
```
Comment (sirpy) [inline .env.example]: same no values

File: .env.test [hunk 7]
```diff
@@ -29,3 +29,5 @@ SKIP_FACE_RECO=false
 ALLOW_FACE_RECO_DUPS=false
 NEW_RELIC_LICENSE_KEY=ab380edbf4e6210529f4aa2513445e7f75672594
 JWT_PASS=G00DDAPP
+WEB3_SITE_URL=https://w3.gooddollar.org
+SECURE_KEY=d41d8cd98f00b204e9800998ecf8427e
```
Comment (sirpy) [inline .env.test]: you can add values to .env.dev

File: src/server/storage/storageAPI.js [hunk 8]
```diff
@@ -42,20 +44,63 @@ const setup = (app: Router, storage: StorageAPI) => {
...
+        process.env.NODE_ENV === 'development'
+          ? Promise.resolve({})
+          : Mautic.createContact(user).catch(e => {
+              log.error('Create Mautic Record Failed', e)
+            })
+
+      const secureHash = md5(`${user.email}${conf.secure_key}`)
+      const web3RecordPromise = user.w3Token
+        ? Promise.rewolve()
+        : fetch(`${conf.web3SiteUrl}/api/wl/user`, {
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: now that we do this for all users, what's going to happen for users that came from W3? ie email is already on W3?

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale please go over all conversation, you missed one last time

File: src/server/storage/storageAPI.js (hunk 8 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: There is a condition for web3RecordPromise to not do this for users that came with web3 token

Comment (sirpy) [reply src/server/storage/storageAPI.js]: when is this w3Token saved to the user's profile?
when is it saved when a user came from W3?

Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: It is saved when the user came from web3 with token

**Comment by sirpy** [comment]
I don't think so. I think you are mixing up the user profile on the dapp
and the user record on the server, they are not the same.

Hadar Rottenberg
Technical Lead, GoodDollar
+972-507 319 093  (telegram <http://t.me/sirpy>, whatsapp)  blog
<https://medium.com/dablock>  gooddollar.org <http://www.gooddollar.org/>



* <https://www.facebook.com/TheGoodDollar>
<https://twitter.com/TheGoodDollar>   <http://instagram.com/good.dollar>
<http://t.me/GoodDollarX>
<https://www.linkedin.com/company/gooddollar/>
<http://medium.com/@GoodDollar>
<http://youtube.com/channel/UC_8oVeHIUo9U8gAc2BwaviA>
<http://reddit.com/user/GoodDollar>   <https://github.com/gooddollar>
  Join the OpenUBI <https://t.me/openUBI> movement!*

File: src/server/storage/storageAPI.js (hunk 8 above)
Comment (sirpy) [reply src/server/storage/storageAPI.js]: iI don't think so. I think you are mixing up the user profile on the dapp and the user record on the server, they are not the same.

Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: The web3 token saved to server user record while there is /user/add request initiated. So the web3 token present in that request payload

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale resolve conflicts

**Comment by sirpy** [comment]
@StanislavShevchenko waiting your review

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale build+coverage not passing

File: .env.dev [hunk 9]
```diff
@@ -23,4 +23,6 @@ SKIP_EMAIL_VERIFICATION=true
 SKIP_FACE_RECO=false
 ALLOW_FACE_RECO_DUPS=true
 ROLLBAR_TOKEN=
+WEB3_SITE_URL=https://w3.gooddollar.org
+SECURE_KEY=d41d8cd98f00b204e9800998ecf8427e
```
[the remark this answers is not in the document]
Comment (yaroslav-fedyshyn-nordwhale) [reply .env.dev]: Fixed in new commit

**Comment by yaroslav-fedyshyn-nordwhale** [comment]
Conflicts resolved

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale resovle and merge

**Pull request merged**